lib/elastic_apm/context.rb

```diff
@@ -0,0 +1,136 @@
+# frozen_string_literal: true
+
+module ElasticAPM
+  # @api private
+  class Context
+    # @api private
+    class Request
+      SKIPPED_PORTS = {
+        'http' => 80,
+        'https' => 443
+      }.freeze
+
+      def initialize
+        @socket = {}
+        @headers = {}
+        @cookies = {}
+        @env = {}
+      end
+
+      attr_accessor :body, :cookies, :env, :headers, :http_version, :method,
+        :socket, :url
+
+      # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+      def add_rack_env(rack_env)
```

Comment:
I find the handling in `add_rack_env` relatively confusing as I cannot find a consistent usage of virtual attrs, initializing hashes in the init method, accessing and setting attrs, etc. 

If `add_rack_env` would be called twice on the same object, then the information in `socket` would be set according to the second `req`, but `headers` would be basically merged from the first and the second `req`. This is an outcome of once overwriting `socket` completely, but only adding key-values to `headers`, while both initializing in the init method. 
I know this is theoretical, but it shows how different you are initializing and accessing the attrs here. 

Maybe you find a way to make this more consistent.

Reply:
Thanks @simitt. You are completely right that this is a bit confusing. I'll try and refactor it to make more sense.